### PR TITLE
feat: add required characters password strength check

### PR DESCRIPTION
**Pull request by hf**
Adds the `GOTRUE_PASSWORD_REQUIRED_CHARACTERS` config option, which if set, will reject passwords that do not contain at least one character of each set of characters.

It is defined like so: `abc...xyz:0123...89`. This means that at least one lowercase and one digit has to be present in the password to be accepted. All other characters are also allowed. To include the `:` character, escape it with `\:`.

When a weak password is detected, the HTTP 429 error is sent with an additional JSON field `weak_password` that includes a `reasons` property -- an array of the strings:

- `length` if the password is not long enough
- `characters` if the password does not use all required character sets

### CODE REVIEW

File: internal/conf/configuration.go [hunk 1]
```diff
@@ -131,6 +131,41 @@ func (c *SessionsConfiguration) Validate() error {
 	return nil
 }
 
+type PasswordRequiredCharacters []string
+
+func (v *PasswordRequiredCharacters) Decode(value string) error {
+	parts := strings.Split(value, ":")
+
+	for i := 0; i < len(parts)-1; i += 1 {
```
Comment (kangmingtay) [inline internal/conf/configuration.go:139]: you'll miss the last character in the string 
```suggestion
	for i := 0; i <= len(parts)-1; i += 1 {
```

File: internal/conf/configuration.go [hunk 2]
```diff
@@ -131,6 +131,41 @@ func (c *SessionsConfiguration) Validate() error {
...
+		if part != "" {
+			*v = append(*v, part)
+		}
+	}
+
+	return nil
+}
+
+type PasswordConfiguration struct {
+	MinLength int `josn:"min_length" split_words:"true"`
```
Comment (kangmingtay) [inline internal/conf/configuration.go:164]:
```suggestion
	MinLength int `json:"min_length" split_words:"true"`
```

File: internal/conf/configuration.go [hunk 3]
```diff
@@ -131,6 +131,41 @@ func (c *SessionsConfiguration) Validate() error {
...
+
+	RequiredCharacters PasswordRequiredCharacters `json:"required_characters" split_words:"true"`
```
Comment (kangmingtay) [inline internal/conf/configuration.go:166]: hmm i'm not too clear how this is defined even after reading the description in the PR - can you give more examples in the README that cover the following cases:
* password must have at least 1 special character
* password must have at least 1 uppercase character
* password must have at least 1 number

File: internal/api/password.go [hunk 4]
```diff
@@ -1,12 +1,49 @@
...
+	if len(password) < config.Password.MinLength {
+		reasons = append(reasons, "length")
+		messages = append(messages, fmt.Sprintf("Password should be at least %d characters.", config.Password.MinLength))
+	}
+
+	for _, characterSet := range config.Password.RequiredCharacters {
+		if characterSet != "" && !strings.ContainsAny(password, characterSet) {
+			reasons = append(reasons, "characters")
+
+			messages = append(messages, fmt.Sprintf("Password should contain at least one character of each: %s.", strings.Join(config.Password.RequiredCharacters, ", ")))
```
Comment (kangmingtay) [inline internal/api/password.go:36]: instead of appending multiple lines of `Password should contain at least one character .... `, can we have this at the top level and then concatenate the missing required characters to the string so that it will just be 1 message?

**Review by kangmingtay — APPROVED** [review]

File: internal/api/password.go (hunk 4 above)
Comment (hf) [reply internal/api/password.go:36]: Actually the `break` at the bottom deals with the problem. It's only 1 message.

File: internal/conf/configuration.go (hunk 1 above)
Comment (hf) [reply internal/conf/configuration.go:139]: Actually this is intentional, the last "part" is always going to be the proper set. It also works because of the `i+1` logic below, which would panic.

Comment (hf) [reply internal/conf/configuration.go:139]: The tests should handle all edge cases so I think it's safe.

File: internal/conf/configuration.go (hunk 3 above)
Comment (hf) [reply internal/conf/configuration.go:166]: It depends on what you set as character sets here. For some applications only long numeric passwords will work, while others may ask for non-English passwords.
